staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -629,3 +637,53 @@ func isTooLargeResourceVersionError(err error) bool {
 	}
 	return false
 }
+
+type RetryWithDeadline interface {
+	Retry() bool
+}
+
+type retryWithDeadlineImpl struct {
+	firstErrorTime   time.Time
+	lastErrorTime    time.Time
+	maxRetryDuration time.Duration
+	minResetPeriod   time.Duration
```

Comment:
Do we really need a concept of minResetPeriod?

I would assume that we should let the retry-er know about every single result of watchHandler and basically any other (different than what we want to retry) error is erasing the previous error (and restarts the counter). That also removes the need for firstError and lastError (only the first is needed, right)?

Reply:
In case of watchcache I don't think it's possible to have IsTooManyRequests error, but I can imagine that if we would have got IsTooManyRequests error we shouldn't erase "no leader" as it could potentially mean that the request didn't go through. So I prefer to have retrier keep track only of internal errors and in this case I need firstError to start retry timer for "no leader" and also "lastError" to estimate time it "worked" so we can reset retry timer.